packages/phone-number-privacy/signer/src/signer/domain/sign.action.ts

```diff
@@ -0,0 +1,105 @@
+import {
+  domainHash,
+  DomainRestrictedSignatureRequest,
+  ErrorMessage,
+  WarningMessage,
+} from '@celo/phone-number-privacy-common'
+import { EIP712Optional } from '@celo/utils/lib/sign-typed-data-utils'
+import { Config } from '../../config'
+import { getDatabase } from '../../database/database'
+import { DomainStateRecord } from '../../database/models/domainState'
+import { DefaultKeyName, Key } from '../../key-management/key-provider-base'
+import { SignAbstract } from '../sign.abstract'
+import { DomainQuotaService } from './quota.service'
+import { DomainSession } from './session'
+import { DomainSignIO } from './sign.io'
+
+export class DomainSignAction extends SignAbstract<DomainRestrictedSignatureRequest> {
+  constructor(
+    readonly config: Config,
+    readonly quota: DomainQuotaService,
+    readonly io: DomainSignIO
+  ) {
+    super()
+  }
+
+  public async perform(session: DomainSession<DomainRestrictedSignatureRequest>): Promise<void> {
+    const domain = session.request.body.domain
+    session.logger.info('Processing request to get domain signature ', {
+      name: domain.name,
+      version: domain.version,
+      hash: domainHash(domain).toString('hex'),
+    })
+
+    try {
+      await getDatabase().transaction(async (trx) => {
+        // Get the current domain state record, or use an empty record if one does not exist.
+        const domainStateRecord = await this.quota.getQuotaStatus(session, trx)
+
+        if (!this.nonceCheck(domainStateRecord, session)) {
+          return this.io.sendFailure(
+            WarningMessage.INVALID_NONCE,
+            401,
+            session.response,
+            domainStateRecord.toSequentialDelayDomainState()
+          )
+        }
+
+        const quotaStatus = await this.quota.checkAndUpdateQuotaStatus(
+          domainStateRecord,
+          session,
+          trx
+        )
+
+        if (!quotaStatus.sufficient) {
+          session.logger.warn(`Exceeded quota`, {
+            name: domain.name,
+            version: domain.version,
+            hash: domainHash(domain),
+          })
+          return this.io.sendFailure(
+            WarningMessage.EXCEEDED_QUOTA,
+            429,
+            session.response,
+            quotaStatus.state.toSequentialDelayDomainState()
+          )
+        }
+
+        const defaultKey: Key = {
+          version: this.config.keystore.keys.domains.latest,
+          name: DefaultKeyName.DOMAINS,
+        }
+
+        // Compute signature inside transaction so it will rollback on error.
+        const { signature, key } = await this.sign(
+          session.request.body.blindedMessage,
+          defaultKey,
+          session
+        )
+
+        this.io.sendSuccess(
+          200,
+          session.response,
+          key,
+          signature,
+          quotaStatus.state.toSequentialDelayDomainState()
+        )
+      })
```

Comment:
Although the signing should occur in the transaction, sending the response to the client should not. Two reasons I say this are:
* Sending the response to the user may take a while time and we want to hold a lock on the database for as little time as possible. In a pathologically case, a client that is slow to receive data, possibly intentionally, could could this operation to take abnormally long as part of a DoS.
* It's possible a client may capable of causing the response to fail after it receives the signature (e.g. by refusing to send an `ACK` for the response data and causing a TCP timeout). If such a case results in the database write being rolled back, then they are able to get a (P)OPRF evaluation without paying any quota which is a security violation.
```suggestion
      })
      this.io.sendSuccess(
          200,
          session.response,
          key,
          signature,
          quotaStatus.state.toSequentialDelayDomainState()
        )
```

Reply:
That makes sense, but this change actually breaks the code because signature, key and quotaStatus are defined inside the transaction. I had to do something kinda whacky to get this to work without using 'let' and compromising type safety. It's a little hacky now but might lead to a cleaner response handling scheme if I work into throughout the code a bit. In the interest of time I'll punt on that for now but come back to it